crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -367,136 +404,184 @@ pub(super) struct FlatAbi {
     pub(super) align: u32,
 }
 
-/// Represents a pending event on a host-owned write end of a stream or future.
-///
-/// See `ComponentInstance::start_write_event_loop` for details.
-enum WriteEvent<B> {
-    /// Write the items in the specified buffer to the stream or future, and
-    /// return the result via the specified `Sender`.
-    Write {
-        buffer: B,
-        tx: oneshot::Sender<HostResult<B>>,
-    },
-    /// Drop the write end of the stream or future.
-    Drop(Option<Box<dyn FnOnce() -> B + Send + Sync>>),
-    /// Watch the read (i.e. opposite) end of this stream or future, dropping
-    /// the specified sender when it is dropped.
-    Watch { tx: oneshot::Sender<()> },
+/// Trait representing objects (such as streams, futures, or structs containing
+/// them) which require access to the store in order to be disposed of properly.
+pub trait DropWithStore: Sized {
+    /// Dispose of `self` using the specified store.
+    fn drop(self, store: impl AsContextMut) -> Result<()>;
+
+    /// Dispose of `self` using the specified accessor.
+    fn drop_with(self, accessor: impl AsAccessor) -> Result<()> {
+        accessor.as_accessor().with(move |store| self.drop(store))
+    }
 }
 
-/// Represents a pending event on a host-owned read end of a stream or future.
-///
-/// See `ComponentInstance::start_read_event_loop` for details.
-enum ReadEvent<B> {
-    /// Read as many items as the specified buffer will hold from the stream or
-    /// future, and return the result via the specified `Sender`.
-    Read {
-        buffer: B,
-        tx: oneshot::Sender<HostResult<B>>,
-    },
-    /// Drop the read end of the stream or future.
-    Drop,
-    /// Watch the write (i.e. opposite) end of this stream or future, dropping
-    /// the specified sender when it is dropped.
-    Watch { tx: oneshot::Sender<()> },
+/// Trait representing objects (such as the writable end of a future or a struct
+/// containing one) which require access to the store _and_ a value to write in
+/// order to be disposed of properly.
+pub trait DropWithStoreAndValue<T: func::Lower + Send + Sync + 'static>: Sized {
```

Comment:
In the abstract I think I personally prefer to keep the closure-to-manufacture-the-value in the constructor of a future rather than having this trait. Would that still work?